### PR TITLE
Modify

### PR DESCRIPTION
**Pull request by DaHaiHuha**
finished the modified work

### CODE REVIEW

**Comment by joinssmith** [comment]
@DaHaiHuha Hi, Thank you for your work. But there are a few questions.  
  
1. As mentioned here[here](https://github.com/sshaoshuai/PointRCNN/issues/110), you seem to have made no changes?  
> For anyone interested, the distance_based_proposal method https://github.com/sshaoshuai/PointRCNN/blob/master/lib/rpn/proposal_layer.py#L58 has to be updated. Such that negative values should be added to nms_range_list since it checks z in the cam dimension. I was able to get behind (negative z)detections afterward.  
  
2. https://github.com/DaHaiHuha/PointRCNN/blob/modify/lib/datasets/lyft_dataset.py#L170, t seems that only the label of RPN is given. Hasn't RCNN written yet?  
  
3. https://github.com/DaHaiHuha/PointRCNN/blob/modify/lib/datasets/lyft_dataset.py#L89 Lyft point cloud coordinate system without conversion, the depth should be ```pts_rect[:, 1]  # (x, y, z, ...)```? That means the Y-axis represents depth.  [paper](https://arxiv.org/abs/1903.11027)

**Comment by DaHaiHuha** [comment]
@joinssmith Well, Thank you very much for your attention and suggestions!

1. Actually, I made this pull request by mistake and this fork is full of bugs and issues unfinished still, as you mentioned above, I do encounter this problem.
image
But, now I'd like to try another way: Modify more on the format of the dataset, obviously, which is not so clever. 
2. You're true, while if we train 'RPN' & 'RCNN' separately, this function is enough.
3. Surely, this is a huge hole of bugs, I use codes below instead:
`        lidar_pc = lidar_pc[:, [1, 2, 0]]
        lidar_pc[:, 0] = -lidar_pc[:, 0]
        lidar_pc[:, 1] = -lidar_pc[:, 1]`

**Comment by DaHaiHuha** [comment]
Plus one, I don't think lyft coordinate is the same as NuSscenes, one is x pointing front and one is y pointing front.
Here are code snipaste from lyft-devkit
```
        # 3D bounding box corners. (Convention: x points forward, y to the left, z up.)
        x_corners = length / 2 * np.array([1, 1, 1, 1, -1, -1, -1, -1])
        y_corners = width / 2 * np.array([1, -1, -1, 1, 1, -1, -1, 1])
        z_corners = height / 2 * np.array([1, 1, -1, -1, 1, 1, -1, -1])
```

**Comment by joinssmith** [comment]
>   You're true, while if we train 'RPN' & 'RCNN' separately, this function is enough.  

I understand it as train RCNN network with fixed RPN network. If that were so, I found that you turned off ROI_SAMPLE_JIT so that ```input_data``` [here](https://github.com/DaHaiHuha/PointRCNN/blob/modify/lib/net/rcnn_net.py#L160) could hardly correspond to [here](https://github.com/DaHaiHuha/PointRCNN/blob/modify/lib/net/point_rcnn.py#L71)  ```rcnn_input_info```.  
  
And I think the cls_label in get_loss_rcnn should be a multi-category label(including all lyft categories), it's not provided in the dataloader, so where is it when ROI_SAMPLE_JIT is turned off?  
  
I'm sorry I can't run your code directly(It's not your problem), I have no way to check whether my view is correct, so it raises some strange questions.

**Comment by DaHaiHuha** [comment]
@joinssmith  Thanks for your reply. Still i have a question, so we have to set ROI_SAMPLE_JIT to be true or we could not train a multi-class model online? Very confused. Anyway, I am working on it

**Comment by joinssmith** [comment]
@DaHaiHuha About multi-class labeling, I did set ROI_SAMPLE_JIT bo be True and pass a 'cls_labels' in the dataloader return dict.  
  
In any case, I find the author commenting [here](https://github.com/sshaoshuai/PointRCNN/blob/master/lib/rpn/proposal_target_layer.py#L78) ```param gt_boxes3d: (B, N, 8) [x, y, z, h, w, l, ry, cls]```. But, I found that gt_boxes3d.shape = (B, N,7), so I think I have to pass in a multi-category tag myself and replace the code [here](https://github.com/sshaoshuai/PointRCNN/blob/master/lib/rpn/proposal_target_layer.py#L63). Otherwise, ```batch_cls_label = (batch_roi_iou > cfg.rcnn. CLS_FG_THRESH). Long ()```only marks the FG rather than the specific category.  
  
I'd like to know your opinion.

**Comment by DaHaiHuha** [comment]
@joinssmith Thanks for your well-rounded reply, I've been solving this problem these days and just as what you said, the key to multi-class labeling for `ROI_SAMPLE_JIT=True` is to pass a gt via data loader. So I've added one and passed it to  `sample_rois_for_rcnn ` alongside the `batch_roi_iou`, so I could directly assign a label for the points. The modified code will be pushed within several days. Now I have to test its performance. Thanks!

**Comment by muzi2045** [comment]
Hi, which train setting file  are you using? @DaHaiHuha 
but train it on 1080ti within 5 epochs, the compute rpn_loss is between 3.x ~ 5.x.
the rpn loss is normally? thanks for any reply.

```
TRAIN:
    SPLIT: train
    VAL_SPLIT: smallval

    LR: 0.002
    LR_CLIP: 0.00001
    LR_DECAY: 0.5
    DECAY_STEP_LIST: [50, 100, 150, 180, 200]
    LR_WARMUP: False
    WARMUP_MIN: 0.0002
    WARMUP_EPOCH: 5

    BN_MOMENTUM: 0.9
    BN_DECAY: 0.5
    BNM_CLIP: 0.01
    BN_DECAY_STEP_LIST: [50, 100, 150, 180, 200]

    OPTIMIZER: adam  # adam, adam_onecycle
    WEIGHT_DECAY: 0.001  # L2 regularization
    MOMENTUM: 0.9

    MOMS: [0.95, 0.85]
    DIV_FACTOR: 10.0
    PCT_START: 0.4

    GRAD_NORM_CLIP: 1.0

    RPN_PRE_NMS_TOP_N: 12000
    RPN_POST_NMS_TOP_N: 2048
    RPN_NMS_THRESH: 0.85
    RPN_DISTANCE_BASED_PROPOSE: True
```
